src/main/java/co/cask/gcp/publisher/source/GoogleSubscriber.java

```diff
@@ -0,0 +1,109 @@
+import org.apache.spark.streaming.pubsub.SparkGCPCredentials;
+import org.apache.spark.streaming.pubsub.SparkGCPCredentials.Builder;
+import org.apache.spark.streaming.pubsub.SparkPubsubMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+/**
+ * Realtime source plugin to read from Google PubSub.
+ */
+@Plugin(type = StreamingSource.PLUGIN_TYPE)
+@Name("GoogleSubscriber")
+@Description("Streaming Source to read messages from Google PubSub.")
+public class GoogleSubscriber extends StreamingSource<StructuredRecord> {
+  private static final Logger LOG = LoggerFactory.getLogger(GoogleSubscriber.class);
+  private static final Schema DEFAULT_SCHEMA =
+    Schema.recordOf("event",
+                    Schema.Field.of("message", Schema.of(Schema.Type.STRING)),
+                    Schema.Field.of("id", Schema.of(Schema.Type.STRING)),
+                    Schema.Field.of("timestamp", Schema.of(Schema.Type.STRING))
+    );
+
+  private SubscriberConfig config;
+
+  public GoogleSubscriber(SubscriberConfig conf) {
+    this.config = conf;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) throws IllegalArgumentException {
+    super.configurePipeline(pipelineConfigurer);
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(DEFAULT_SCHEMA);
+  }
+
+  @Override
+  public JavaDStream<StructuredRecord> getStream(StreamingContext streamingContext) {
+    String serviceAccountFilePath = config.getServiceAccountFilePath();
+    SparkGCPCredentials credentials;
+    if (serviceAccountFilePath != null) {
+      // if credential file is provided then use it to build SparkGCPCredentials
+      LOG.debug("Building credentials from specified service account file.");
+      credentials = new Builder().jsonServiceAccount(serviceAccountFilePath).build();
+    } else {
+      // this will use the application default credential which is available when running on google cloud
+      LOG.debug("No service account file specified. Building credentials using Application Default Credentials.");
+      credentials = new Builder().build();
+    }
+
+    JavaReceiverInputDStream<SparkPubsubMessage> pubSubMessages =
+      PubsubUtils.createStream(streamingContext.getSparkStreamingContext(), config.getProject(), config.topic,
+                               config.subscriber, credentials, StorageLevel.MEMORY_ONLY());
+
+    return pubSubMessages.map((Function<SparkPubsubMessage, StructuredRecord>) pubSubMessage ->
+      StructuredRecord.builder(DEFAULT_SCHEMA)
+        .set("message", Bytes.toString(pubSubMessage.getData()))
+        .set("id", pubSubMessage.getMessageId())
+        .set("timestamp", pubSubMessage.getPublishTime()).build());
+  }
+
+
+  public static class SubscriberConfig extends GCPReferenceSourceConfig {
+    @Name("subscriber")
+    @Description("Cloud Pub/Sub subscriber to subscribe. If a subscriber does not exists and needs to be create then " +
```

Comment:
Cloud Pub/Sub subscription to read from. If the subscription does not exist it will be automatically created if a topic is specified. Messages published before the subscription was created will not be read.